test/test_indexing.py

```diff
@@ -649,6 +649,302 @@ def test_advancedindex_big(self, device):
+
+            # https://github.com/pytorch/pytorch/issues/71673
+            # check 0: compositionality
+            self.assertEqual(t[1, m, 0], t[1, :, :, 0][m])
+            # check 1: expected result
+            self.assertEqual(t[1, m, 0], r)
+
+            # check 2: numpy works as expected
+            self.assertEqual(nt[1, nm, 0], nr)
+            # check 3: expected result against numpy
+            self.assertEqual(
+                t[1, m, 0],
+                torch.tensor(nr, device=device, dtype=dtype))
+            # check 4: numpy compositionality
+            self.assertEqual(nt[1, nm, 0], nt[1, :, :, 0][nm])
+
+    # one mask tensor (3d)
+    @dtypes(torch.float)
+    def test_advancedindex_dim_1m_3d(self, device, dtype):
+        t = [
+            self._advancedindex_input1,
+            self._advancedindex_input2,
+        ]
+
+        values = (
+            # tensor, mask, result
+            # case 0: everything
+            (t,
+
+             [[[1., 1., 1.], [1., 1., 1.], [1., 1., 1.], [1., 1., 1.]],
+              [[1., 1., 1.], [1., 1., 1.], [1., 1., 1.], [1., 1., 1.]]],
+
+             # 0  1   2   3   4   5    6    7    8    9    10   11
+             [0., 2., 4., 6., 8., 10., 12., 14., 16., 18., 20., 22.,
+              # 12 13   14   15   16   17   18   19   20   21   22   23
+              24., 26., 28., 30., 32., 34., 36., 38., 40., 42., 44., 46.]),
+
+            # case 1: nothing
+            (t,
+
+             [[[0., 0., 0.], [0., 0., 0.], [0., 0., 0.], [0., 0., 0.]],
+              [[0., 0., 0.], [0., 0., 0.], [0., 0., 0.], [0., 0., 0.]]],
+
+             []),
+
+            # case 2: random
+            (t,
+
+             [[[1., 0., 1.], [0., 1., 1.], [0., 0., 0.], [1., 1., 1.]],
+              [[0., 1., 0.], [1., 1., 0.], [0., 0., 1.], [0., 0., 0.]]],
+
+             [0., 4., 8., 10., 18., 20., 22., 26., 30., 32., 40.]),
+        )
+
+        for tv, mv, rv in values:
+            t = torch.tensor(tv, device=device, dtype=dtype)
+            m = torch.tensor(mv, device=device, dtype=dtype) > 0
+            r = torch.tensor(rv, device=device, dtype=dtype)
+
+            nt = np.array(tv)
+            nm = np.array(mv) > 0
+            nr = np.array(rv)
+
+            # https://github.com/pytorch/pytorch/issues/71673
+            # check 0: compositionality
+            self.assertEqual(t[m, 0], t[:, :, :, 0][m])
+            # check 1: expected result
+            self.assertEqual(t[m, 0], r)
+
+            # check 2: numpy works as expected
+            self.assertEqual(nt[nm, 0], nr)
+            # check 3: expected result against numpy
+            self.assertEqual(
+                t[m, 0],
+                torch.tensor(nr, device=device, dtype=dtype))
+            # check 4: numpy compositionality
+            self.assertEqual(nt[nm, 0], nt[:, :, :, 0][nm])
+
+    # two mask tensors (1d)
+    @dtypes(torch.float)
+    def test_advancedindex_dim_2m_1d(self, device, dtype):
+        t = [
```

Comment:
Can we just unite all the tests into one? They seem to pretty similar. You can write a function that maps inputs to the indexing pattern per test and/or case.

Reply:
they are similar, but the shapes and view operations are different (e.g., `[:,:,...]` vs `[:,...]`), the number of tensors is different as well (`[m]` vs `[m,n]`). it would essentially require special-casing anyway (an `if` for each shape), so imo it'd only complicate things, making it harder to reason about and debug

Reply:
No need to if-else. You just create functions that map inputs to indixed inputs.

Reply:
i have no strong opinion and can do whatever, but i still fail to see the benefits of the above. what are they? rn we have 4 distinct functions or so, but at least each of them fits on the screen when reading thru the code. you can also test separate cases without touching the testsuite at all. and there isn't really any code sharing possible as far as i can tell. so what you're proposing is to just merge all them into a single mega-function, which will be huge. so what's the point again? please carefully look at the cases again, there's no way to share much code there due to the number of `:` being different, or parameter placement (tensor index is the first one, tensor index is the second one), or the number of tensor masks.

again, i'm happy to go with whatever. could you maybe show how you'd do it for `test_advancedindex_dim_1m_2d` and `test_advancedindex_dim_2m_1d`?

Reply:
also, as a heads up: not a fan of zipping due to its (intended) behavior. it stops processing input once you reach the end of one of the arguments. this is error-prone and could result in certain things being left out, which can happen eventually when more things are added to the testsuite, unless care is taken to make sure that the args have the same size:

```py
>>> list(zip([1,2], ['a','b','c','d']))
[(1, 'a'), (2, 'b')]
```

Reply:
As far as I see the processing code is pretty much the same. What distinguishes a test is a tuple (input, mask, collection of functions that define indexing cases). It is up to you, but if new cases arise, we would need to copy-paste the entire body of a test. I am also a bit confused about the name, as it is not clear as to what `dim_2m_1d` is supposed to test. Maybe a comment would not hurt?

Reply:
regarding naming:

```py
# two mask tensors (1d)
@dtypes(torch.float)
def test_advancedindex_dim_2m_1d
```

- `dim` -- we are testing the dimension pointer offset thing (aka the cause of the bug). this also allows triggering only this set of `advancedindex` tests -- there are other ones, which were added before
- `2m` -- 2 mask tensors (mentioned in the comment above the test, need to check whether it breaks if there are multiple tensor index args)
- `1d` -- tensors are 1-dimensional (need to check 1d and >1d cases).

yes, i think copy-pasting in this case is fine. i think this way it's much more clear what case is actually being tested. i'll add more comments regarding what each test does and why tho.

Reply:
i'd also like to point out that i link to the relevant issue everywhere, so i'd expect people to read thru it before touching any tests here. and the issue includes extensive discussion regarding how indexes work as well as a pointer to the dev-discuss post by ed.

Reply:
@nikitaved added more comments to the tests, lmk if there are still issues with this